### DISCUSSION TITLE
CUDA-aware support

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#26798] [Q&A Getting Started]
Dear all,
quick question: I recently installed MOOSE (+ our applications) on the leonardo booster module at CINECA in Italy. All went fine and smooth, but when trying to run the tests  (failing) -  errors are related to having disabled the CUDA-aware support (this is by deqfault required from the openMPI module installed locally). This said, I was wondering if there is a simple way to disable the CUDA-aware support at the TestHarness script level, or in other words, where I can integrate the --mca  opal_warn_on_missing_libcuda 0 option. If not, I might go all the way to install locally openmpi. Please, keep in mind that this is not a top priority discussion, since production runs are fine (at least as long as I have tested), and it is only because I found it annoying not to be able to run the tests upon a new installation (users keep on asking me why tests are failing...). 
Thanks for any advise, 
mauro

**Comment by milljm** [comment]
If I am reading you correctly, you may want to use:

```bash
export MOOSE_MPI_COMMAND='mpiexec --opal_warn_on_missing_libcuda 0'
```
This _might_ work, if all you need is a way to add some arguments to `mpiexec`.

If instead you need to pass arguments to your MOOSE based application I believe you're after the `--cli-args` TestHarness option:
```pre
./run_tests --help

<trimmed>
  --cli-args [CLI_ARGS] Append the following list of arguments to the 
                        command line (Encapsulate the command in quotes)
```


I was looking for a decent doco page on our TestHarness, but this was all I could find: https://mooseframework.inl.gov/python/TestHarness.html

TestHarness options/influential-environment-variables documentation could really use some TLC... https://github.com/idaholab/moose/issues/26799

**Reply by mcacace** [reply]
dear @milljm thanks a lot for your answer! It seems that adding the flag to the mpiexec works, at least against the CUDA aware support. However, all tests are crashing (signal 11 segmentation fault on one PID). Not sure whether this has to do with the cluster or not, but as said, production runs are fine.

**Reply by milljm** [reply]
I am not sure if @grmnptr can help, but, well 'CUDA'. Any ideas?

It sounds like the TestHarness is botching things up somehow, if we're able to manually run the application against the same test(s) without issue.

**Reply by mcacace** [reply]
I come back to this issue, though the problem slightly changed (I can open a new issue in case). However, about the test failing if based on the run_test script.... I noticed that upon a rebase of the conda environment, I keep receiving some warnings about VTKMPIController being already initialized at the beginning of each runs as well as some other related to SlepcInitialize() being called after SlepcFinalize() at the end of each runs. Could those be the reasons why the run_test script is crashing?

**Reply by mcacace** [reply]
I mean at least for conda based installation. The issue with the HPC, still remains, but there I am still without a clue. 
Thanks for any feedback
mauro

**Reply by GiudGiud** [reply]
Can you please paste the entire error you see?
As well as:
- the output of ldd on the executable
- the output of the diagnostics script in moose/scripts

**Reply by mcacace** [reply]
dear @GiudGiud here are the required files. 
[diagnostic_moose.log](https://github.com/idaholab/moose/files/14347873/diagnostic_moose.log)
[list_moose.log](https://github.com/idaholab/moose/files/14347882/list_moose.log)
[error.log](https://github.com/idaholab/moose/files/14347969/error.log)
About the errors, the one on set up of a run is as follows (it repeats per the number of processors on the screen only, not sure why, but could not dump in a log file)
2024-02-20 16:48:15.488 (   0.833s) [        6DD9CC80]   vtkMPIController.cxx:136   WARN| vtkMPIController (0x55e368e8f0f0): Already initialized.

The error at the end of a run, you can find in the log file (attached)

Thanks a lot for helping,
mauro

**Reply by GiudGiud** [reply]
Hello

instead of the conda package list I would like the output of `ldd <name of your executable-opt>`

**Reply by GiudGiud** [reply]
Is this the simulation you care about btw? Because it s very small for using GPUs or even multiple CPUs

**Reply by mcacace** [reply]
It was a ldd on my executable (golem-opt), but was missing to activate the moose environment (sorry about that).  Here the update file:
[golem.log](https://github.com/idaholab/moose/files/14348540/golem.log)
About the test run, it is a very small example (just single timestep) only to showcase the warnings I am getting. So, no real production runs, though the relevant warnings are the same. 
Thanks,
Mauro

**Reply by mcacace** [reply]
If I might add something, I had to reinstall all from scratch on another cluster, because of an upgrade of the system. All went smooth, all moose tests are passing but those from my app are still crashing (cannot have any informative hint, apart reason crash though). I also noticed that while compiling the system complains on NEML and libtorch missing, but I think this is just a warning. And again, all simulations do run fine if launched with proper slurm commands.

**Reply by GiudGiud** [reply]
ah thanks that s good information.
Can you save all your work then `git clean -xfd` your app (and maybe moose too)
then rebuild moose and your app

my concern is that there are some left over compiled objects from before the system was upgraded

**Reply by GiudGiud** [reply]
I dont see anything wrong in the diagnostics. Though do we expect more cuda libraries in that log @grmnptr ?

**Reply by mcacace** [reply]
Sorry, but I think I attached the wrong files (those read to me as my local conda installation). I'm out of office until tomorrow, and will double check. SORRY

**Reply by GiudGiud** [reply]
Ok no worries. 
Let s look into it when you have them

**Comment by mcacace** [comment]
So I think I managed to solve the issue. It all started with me trying to get rid of this warning (after updating the framework)
"*** Warning, This code is deprecated and will be removed in future versions:
Please update your main.C to adapt new main function in MOOSE framework, see 'test/src/main.C in MOOSE as an example of moose::main()'."

For which I thought it smart to update my main.C as follows:

From the old version: 
// Create an instance of the application and store it in a smart pointer for easy cleanup
`std::shared_ptr<MooseApp> app = AppFactory::createAppShared("GolemApp", argc, argv);`
// Execute the application
`app->run();`

to the new version:

`Moose::main<GolemApp>(argc, argv);`

That creates the error=crashed for all the tests. In addition, my fault, I never had a look at the log file from my students runs on the cluster, which, I realized this morning, did run till the end, but throwing an segfault message before finalizing. 

So the question that remains, is what's the right way to update the pointer to the app. Would something like the following work fine with the new syntax?

```
MooseApp * app = AppFactory::createApp("GolemApp", argc, argv);
app->run();
delete app;
```

Apart from that I think we can close the issue, again, sorry for bothering you all that much and thanks for the support.
Mauro

**Reply by mcacace** [reply]
On a side note, the CUDA-aware support issue has been solved by:
`export MOOSE_MPI_COMMAND='mpiexec --opal_warn_on_missing_libcuda 0'`
